### DISCUSSION TITLE
Line numbers in Neo-Tree

### DISCUSSION THREAD

**Discussion by Bogfoot** [nvim-neo-tree/neo-tree.nvim#843] [Q&A] [answered]
Greetings! I'd like to start off by saying that I'm really enjoying neotree and I am thankful to the developers for maintaining and improving it!

My question is related to line numbers in neotree. Is there a way to always show neotree line numbers? 

So far what I was doing when in a larger file tree I would have to go to the neotree buffer and type `:set rnu` and then jump to wherever I needed to go from there. This solution works but clearly the line numbers are lost whenever I close neotree.

Is there a way to make this persistent or to specify a config option to always keep it on? I've searched through the wiki and a bit through the readme today but couldn't find it being mentioned just yet.

**Comment by miversen33** [comment, marked as answer]
After a bit of digging, I figured out where buffer/window options are being set.
[neo-tree.ui.renderer.lua:L831-L834](https://github.com/nvim-neo-tree/neo-tree.nvim/blob/fee76c06e2bfd7d3dbb39b197fc208f6d5c0410c/lua/neo-tree/ui/renderer.lua#L831-L834)

```lua
    win_options = {
      colorcolumn = "",
      signcolumn = "no",
    },
```

These options are passed to [Nui.nvim](https://github.com/MunifTanjim/nui.nvim/blob/5edb8af028f84e0ed3a3999ccb7be4f4a8f7c120/lua/nui/utils/init.lua#L124-L131) where they are applied to the buffer that is created by Nui via Neo-tree.

As you see, there is no (current) way to influence the window options that are generated.

The current Neo-tree state is provided to the function `renderer.create_window` function however, and above in that same function we actually get some config options
```lua
  local default_position = utils.resolve_config_option(state, "window.position", "left")
  local relative = utils.resolve_config_option(state, "window.relative", "editor")
```

In theory this could just be another configuration option that we tie to `window` (something like `window.numberline`?).

I imagine this would look like
```lua
local numberline = utils.resolve_config_option(state, 'window.numberline', false)
-- .. Skipping down to the win_options
win_options = {
-- Everything that is already in here
  -- New variable pulled from config
  number = numberline
}
```

Of course, this is something that @cseickel  would need to be ok with, and likely we would need to convert this to an issue if they decide they _are_ ok with this. Also I could be completely wrong and there is some way to do this that I did not see

**Reply by Bogfoot** [reply]
Thank you very much for your reply.  I'll try to see if there's a way for me to figure it out with the information you've provided!

**Reply by miversen33** [reply]
Just to clarify, my answer is basically "No you cannot right now, unless I missed something". There isn't really a "figure it out" here. The solution is most likely a PR that would need to be created and merged.

Again, this does rely on me being right which I may not be, but the information above is not meant to be used for a configuration you can input or anything, its the information that you would need if you were to submit a PR to resolve this within the neo-tree code.

**Reply by Bogfoot** [reply]
For now I just changed the mentioned lines in `.local/share/nvim/lazy/neo-tree/lua/neo-tree/ui/renderer.lua`. It does what I want so I'm happy! Thank you a lot :) 

![2023-04-05_19:23:26](https://user-images.githubusercontent.com/16871247/230156764-c2ae54a3-e2cb-45a0-b4ec-e903e7814f1d.png)

**Comment by Sebastian-0** [comment]
There is a solution for this already that doesn't require changing the source code, see the answer to this issue: https://github.com/nvim-neo-tree/neo-tree.nvim/issues/608 

Essentially you add this to your config:
```lua
require("neo-tree").setup({
  event_handlers = {
    {
      event = "neo_tree_buffer_enter",
      handler = function(arg)
        vim.opt.relativenumber = true
      end,
    }
  },
```

You can also add other options here, like `vim.opt.number = true` if you also want to see the current line number.

**Reply by ryardley** [reply]
@Bogfoot this should be reselected as the correct answer.

**Reply by Bogfoot** [reply]
Thank you!
